allennlp/modules/seq2vec_encoders/boe_encoder.py

```diff
@@ -0,0 +1,67 @@
+from overrides import overrides
+
+import torch
+from torch.autograd import Variable
+
+from allennlp.common import Params
+from allennlp.modules.seq2vec_encoders.seq2vec_encoder import Seq2VecEncoder
+from allennlp.nn.util import get_lengths_from_binary_sequence_mask
+
+@Seq2VecEncoder.register("boe")
+class BagOfEmbeddingsEncoder(Seq2VecEncoder):
+    """
+    A ``BagOfEmbeddingsEncoder`` is a simple :class:`Seq2VecEncoder` which simply sums the embeddings
+    of a sequence across the time dimension. The input to this module is of shape ``(batch_size, num_tokens,
+    embedding_dim)``, and the output is of shape ``(batch_size, embedding_dim)``.
+
+    Parameters
+    ----------
+    embedding_dim: ``int``
+        This is the input dimension to the encoder.
+    averaged: ``bool``, optional (default=``False``)
+        If ``True``, this module will average the embeddings across time, rather than simply summing
+        (ie. we will divide the summed embeddings by the length of the sentence).
+    """
+    def __init__(self,
+                 embedding_dim: int,
+                 averaged: bool = False) -> None:
+        super(BagOfEmbeddingsEncoder, self).__init__()
+        self._embedding_dim = embedding_dim
+        self._averaged = averaged
+
+    @overrides
+    def get_input_dim(self) -> int:
+        return self._embedding_dim
+
+    @overrides
+    def get_output_dim(self) -> int:
+        return self._embedding_dim
+
+    def forward(self, tokens: torch.Tensor, mask: torch.Tensor = None):  #pylint: disable=arguments-differ
+        if mask is not None:
+            tokens = tokens * mask.unsqueeze(-1).float()
+
+        # Our input has shape `(batch_size, num_tokens, embedding_dim)`, so we sum out the `num_tokens`
+        # dimension.
+        summed = tokens.sum(1)
+
+        if self._averaged:
+            if mask is not None:
+                lengths = get_lengths_from_binary_sequence_mask(mask)
+            else:
+                lengths = Variable(tokens.data.new().resize_(1).fill_(tokens.size(1)), requires_grad=False)
```

Comment:
I think this line and the one below would be cleaner if you used `torch.cuda.device_of()`: https://github.com/allenai/allennlp/blob/b164061b42380ea98641d3efce2789b3cc18e9d3/allennlp/models/encoder_decoders/simple_seq2seq.py#L138-L139

Reply:
Thanks, I've never seen that pattern before. I just tried it in python and it didn't appear to do what I need (namely declare "lengths" on the GPU if the input is on GPU):

```
>>> x = torch.rand(2,3).cuda()
>>> with torch.cuda.device_of(x): y = torch.rand(2,3)
... 
>>> y

 0.9666  0.9871  0.3451
 0.0808  0.9643  0.7967
[torch.FloatTensor of size 2x3]
```

What am I missing?

Reply:
Hmm...  Looks like @pdasigi's fix doesn't actually work.  I just tried it too.  Ignore this comment, then.  @pdasigi, we need to fix the seq2seq model.